File: backend/src/routes/investments.js
import express from 'express';
import { v4 as uuidv4 } from 'uuid';
import { query, getOne, getMany } from '../db.js';
import { authMiddleware } from '../middleware/auth.js';
import { broadcast } from '../server.js';

const router = express.Router();

// All routes require authentication
router.use(authMiddleware);

// Get all investments
router.get('/', async (req, res) => {
  try {
    const { sort = '-purchase_date' } = req.query;

    let sql = 'SELECT * FROM investments WHERE created_by = $1';
    const params = [req.user.id];

    // Handle sorting
    if (sort) {
      const sortField = sort.startsWith('-') ? sort.substring(1) : sort;
      const sortDirection = sort.startsWith('-') ? 'DESC' : 'ASC';
      sql += ` ORDER BY ${sortField} ${sortDirection}`;
    } else {
      sql += ' ORDER BY purchase_date DESC';
    }

    const investments = await getMany(sql, params);
    res.json(investments);
  } catch (error) {
    console.error('Get investments error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Get single investment
router.get('/:id', async (req, res) => {
  try {
    const investment = await getOne(
      'SELECT * FROM investments WHERE id = $1 AND created_by = $2',
      [req.params.id, req.user.id]
    );

    if (!investment) {
      return res.status(404).json({ error: 'Investment not found' });
    }

    res.json(investment);
  } catch (error) {
    console.error('Get investment error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Create investment
router.post('/', async (req, res) => {
  try {
    const { name, type, initial_amount, current_value, purchase_date, notes } = req.body;

    if (!name || !type || !initial_amount || !purchase_date) {
      return res.status(400).json({ error: 'Missing required fields' });
    }

    const id = uuidv4();
    const now = new Date();

    const investment = await getOne(
      `INSERT INTO investments (id, name, type, initial_amount, current_value, purchase_date, notes, created_by, created_date)
       VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)
       RETURNING *`,
      [id, name, type, initial_amount, current_value || initial_amount, purchase_date, notes || null, req.user.id, now]
    );

    // Broadcast update via WebSocket
    broadcast({ type: 'investment_created', data: investment, userId: req.user.id });

    res.status(201).json(investment);
  } catch (error) {
    console.error('Create investment error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Update investment
router.put('/:id', async (req, res) => {
  try {
    const { name, type, initial_amount, current_value, purchase_date, notes } = req.body;

    // Check if investment exists and belongs to user
    const existing = await getOne(
      'SELECT * FROM investments WHERE id = $1 AND created_by = $2',
      [req.params.id, req.user.id]
    );

    if (!existing) {
      return res.status(404).json({ error: 'Investment not found' });
    }

    const investment = await getOne(
      `UPDATE investments
       SET name = $1, type = $2, initial_amount = $3, current_value = $4, purchase_date = $5, notes = $6
       WHERE id = $7
       RETURNING *`,
      [
        name ?? existing.name,
        type ?? existing.type,
        initial_amount ?? existing.initial_amount,
        current_value ?? existing.current_value,
        purchase_date ?? existing.purchase_date,
        notes ?? existing.notes,
        req.params.id
      ]
    );

    // Broadcast update via WebSocket
    broadcast({ type: 'investment_updated', data: investment, userId: req.user.id });

    res.json(investment);
  } catch (error) {
    console.error('Update investment error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Delete investment
router.delete('/:id', async (req, res) => {
  try {
    // Check if investment exists and belongs to user
    const existing = await getOne(
      'SELECT * FROM investments WHERE id = $1 AND created_by = $2',
      [req.params.id, req.user.id]
    );

    if (!existing) {
      return res.status(404).json({ error: 'Investment not found' });
    }

    await query('DELETE FROM investments WHERE id = $1', [req.params.id]);

    // Broadcast update via WebSocket
    broadcast({ type: 'investment_deleted', data: { id: req.params.id }, userId: req.user.id });

    res.json({ message: 'Investment deleted successfully' });
  } catch (error) {
    console.error('Delete investment error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

export default router;

File: fastapi_app/routes/investments.py
"""
Investment routes.
"""
import uuid
from typing import List
from fastapi import APIRouter, Depends, HTTPException, status, Query
from models import InvestmentCreate, InvestmentUpdate, InvestmentResponse, MessageResponse
from auth import get_current_user, User
from database import query, query_one, execute

router = APIRouter(prefix="/investments", tags=["Investments"])


@router.get("", response_model=List[InvestmentResponse])
async def list_investments(
    sort: str = Query("desc", description="Sort order by purchase date: 'asc' or 'desc'"),
    current_user: User = Depends(get_current_user)
):
    """
    List all investments for the authenticated user.

    - **sort**: Sort by purchase date ('asc' for oldest first, 'desc' for newest first)
    """
    order = "ASC" if sort.lower() == "asc" else "DESC"
    investments = query(
        f"""
        SELECT id, name, type, initial_amount, current_value, purchase_date, notes,
               created_by, created_date, updated_date
        FROM investments
        WHERE created_by = %s
        ORDER BY purchase_date {order}
        """,
        (current_user.id,)
    )
    return [InvestmentResponse(**i) for i in investments]


@router.get("/{investment_id}", response_model=InvestmentResponse)
async def get_investment(
    investment_id: str,
    current_user: User = Depends(get_current_user)
):
    """
    Get a single investment by ID.
    """
    investment = query_one(
        """
        SELECT id, name, type, initial_amount, current_value, purchase_date, notes,
               created_by, created_date, updated_date
        FROM investments
        WHERE id = %s AND created_by = %s
        """,
        (investment_id, current_user.id)
    )

    if not investment:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Investment not found"
        )

    return InvestmentResponse(**investment)


@router.post("", response_model=InvestmentResponse, status_code=status.HTTP_201_CREATED)
async def create_investment(
    investment: InvestmentCreate,
    current_user: User = Depends(get_current_user)
):
    """
    Create a new investment.

    - **name**: Investment name (e.g., 'Bitcoin', 'Apple Stock')
    - **type**: Investment type (e.g., 'crypto', 'stocks', 'bonds', 'real_estate')
    - **initial_amount**: Initial investment amount
    - **current_value**: Current value of the investment
    - **purchase_date**: Date of purchase
    - **notes**: Additional notes (optional)
    """
    investment_id = str(uuid.uuid4())
    result = execute(
        """
        INSERT INTO investments (id, name, type, initial_amount, current_value, purchase_date, notes, created_by)
        VALUES (%s, %s, %s, %s, %s, %s, %s, %s)
        RETURNING *
        """,
        (
            investment_id,
            investment.name,
            investment.type,
            investment.initial_amount,
            investment.current_value,
            investment.purchase_date,
            investment.notes,
            current_user.id
        )
    )

    return InvestmentResponse(**result)


@router.put("/{investment_id}", response_model=InvestmentResponse)
async def update_investment(
    investment_id: str,
    investment: InvestmentUpdate,
    current_user: User = Depends(get_current_user)
):
    """
    Update an existing investment.
    Only provided fields will be updated.
    """
    existing = query_one(
        "SELECT * FROM investments WHERE id = %s AND created_by = %s",
        (investment_id, current_user.id)
    )

    if not existing:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Investment not found"
        )

    # Build update query dynamically
    updates = []
    params = []

    if investment.name is not None:
        updates.append("name = %s")
        params.append(investment.name)

    if investment.type is not None:
        updates.append("type = %s")
        params.append(investment.type)

    if investment.initial_amount is not None:
        updates.append("initial_amount = %s")
        params.append(investment.initial_amount)

    if investment.current_value is not None:
        updates.append("current_value = %s")
        params.append(investment.current_value)

    if investment.purchase_date is not None:
        updates.append("purchase_date = %s")
        params.append(investment.purchase_date)

    if investment.notes is not None:
        updates.append("notes = %s")
        params.append(investment.notes)

    if not updates:
        return InvestmentResponse(**existing)

    updates.append("updated_date = NOW()")
    params.append(investment_id)
    params.append(current_user.id)

    sql = f"UPDATE investments SET {', '.join(updates)} WHERE id = %s AND created_by = %s RETURNING *"
    result = execute(sql, tuple(params))

    return InvestmentResponse(**result)


@router.delete("/{investment_id}", response_model=MessageResponse)
async def delete_investment(
    investment_id: str,
    current_user: User = Depends(get_current_user)
):
    """
    Delete an investment.
    """
    result = execute(
        "DELETE FROM investments WHERE id = %s AND created_by = %s RETURNING id",
        (investment_id, current_user.id)
    )

    if not result:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Investment not found"
        )

    return MessageResponse(message="Investment deleted successfully")